cbmc_viewer/metric_reportt.py

```diff
@@ -0,0 +1,392 @@
+    'alias': VALID_ALIAS_STATS,
+    'memop': VALID_MEMOP_STATS,
+    'array': VALID_ARRAY_CONSTRAINT_STATS,
+    'cbmc': VALID_CBMC_PROOF_STATS,
+    'clause': VALID_CLAUSE_STATS,
+    'runtimeKissat': float,
+    'runtime': VALID_RUNTIME
+    }],required=True)
+
+VALID_SUMMARY = voluptuous.schema_builder.Schema({
+    'summary': VALID_RUNTIME_ANALYSIS_SUMMARY
+    },required=True)
+
+################################################################
+
+class RuntimeAnalysisSummary:
+
+    def __init__(self, proofs, srcdir):
+        self.summary = []
+
+        for proof in proofs:
+            jobname = proof
+            proof_path = os.path.abspath(proof)
+            outdir = os.path.join(proof_path, 'html', 'html')
+
+            byteop = os.path.join(proof_path, 'html', 'json', 'viewer-byteop.json')
+            alias = os.path.join(proof_path, 'html', 'json', 'viewer-alias.json')
+            memop = os.path.join(proof_path, 'html', 'json', 'viewer-memop.json')
+            array = os.path.join(proof_path, 'html', 'json', 'viewer-array.json')
+            result = os.path.join(proof_path, 'html', 'json', 'viewer-result.json')
+            clause = os.path.join(proof_path, 'html', 'json', 'viewer-clause.json')
+            kissat = os.path.join(proof_path, 'logs', 'kissat.txt')
+            
+            self.summary.append({
+                'jobname': jobname,
+                'link': get_link_to_job(outdir),
+                'byteop': get_byteop_stats(byteop, outdir),
+                'alias': get_alias_stats(alias, outdir),
+                'memop': get_memop_stats(memop, outdir),
+                'array': get_array_stats(array, outdir),
+                'cbmc': get_cbmc_proof_stats(result),
+                'clause': get_clause_stats(clause, outdir),
+                'runtimeKissat': get_kissat_runtime(kissat),
+                'runtime': get_cbmc_runtime_stats(result)
+                })
+
+        self.validate()
+
+
+    def __str__(self):
+        """Render the byte op summary as html."""
+        return templates.render_runtime_analysis_summary(self.summary)
+
+    def validate(self):
+        """Validate members of a summary object."""
+
+        return voluptuous.humanize.validate_with_humanized_errors(
+            self.__dict__, VALID_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir=None):
+        """Write the proof summary to a file rendered as html."""
+
+        util.dump(self, filename or "runtime_analysis_report.html", outdir)
+
+################################################################
+# Utility functions to generate runtime analysis summary
+
+def get_link(outdir, metric):
+    return '<a href="{}/{}.html">'.format(outdir, metric)
+
+def get_link_to_job(outdir):
+    return get_link(outdir, 'index')
+
+def fail(msg):
+    """Log failure and raise exception."""
+
+    logging.info(msg)
+    raise UserWarning(msg)
+
+################################################################
```

Comment:
I'm unable to follow the code in the remainder of this file, can we discuss by phone?